src/element-stub.js

```diff
@@ -15,21 +15,83 @@
  */
 
 import {BaseElement} from './base-element';
-
+import {getMode} from '../src/mode';
 
 /** @type {!Array} */
 export const stubbedElements = [];
 
+let ampAdScriptInsertedOrPresent = false;
+
+// For testing purpose only.
+export function resetAdScriptInsertedOrPresent() {
+  ampAdScriptInsertedOrPresent = false;
+}
+
 export class ElementStub extends BaseElement {
   constructor(element) {
     super(element);
+    this.updateAmpAdScriptInfo(element);
     stubbedElements.push(this);
   }
 
+  updateAmpAdScriptInfo(element) {
+    if (isAmpAdScriptRequired(element, this.getWin())) {
+      const ampAdScript = createAmpAdScript(this.getWin());
+      this.getWin().document.head.appendChild(ampAdScript);
+    }
+  }
+
   /** @override */
   isLayoutSupported(unusedLayout) {
     // Always returns true and will eventually call this method on the actual
     // element.
     return true;
   }
 }
+
+// Calculate script url for amp-ad, export only for testing reasons.
+export function calculateAdScriptUrl(path) {
+  let scriptSrc;
+  if (getMode().localDev) {
+    scriptSrc = 'https://cdn.ampproject.org/v0/amp-ad-0.1.js';
+    if (path.indexOf('max') >= 0) {
+      scriptSrc = 'http://localhost:8000/dist/v0/amp-ad-0.1.max.js';
+    } else if (path.indexOf('min') >= 0) {
+      scriptSrc = 'http://localhost:8000/dist/v0/amp-ad-0.1.js';
+    }
+  } else {
+    const domain = 'https://cdn.ampproject.org/';
+    const folderPath = getMode().version == '$internalRuntimeVersion$' ?
+        '' : `rtv/${getMode().version}/`;
+    scriptSrc = `${domain}${folderPath}v0/amp-ad-0.1.js`;
+  }
+  return scriptSrc;
+}
+
+// Create the missing amp-ad HTML script element.
+function createAmpAdScript(win) {
+  const ampAdScript = win.document.createElement('script');
+  ampAdScript.async = true;
+  ampAdScript.setAttribute('custom-element', 'amp-ad');
+  ampAdScript.setAttribute('data-script', 'amp-ad');
+  const pathStr = win.location.pathname;
+  const scriptSrc = calculateAdScriptUrl(pathStr);
+  ampAdScript.src = scriptSrc;
+  return ampAdScript;
+}
+
+// Determine the need to add amp-ad script to document.
+function isAmpAdScriptRequired(element, win) {
```

Comment:
Please update comment as outlined above. We typically pass in the window as the first argument.